trino/client.py

```diff
@@ -925,3 +977,217 @@ def _parse_retry_after_header(retry_after):
+    segment from remote storage.
+
+    Attributes:
+        rows (property): The rows of data, loaded and mapped from the spooled segment.
+        uri (property): The URI for the spooled segment.
+        ack_uri (property): The URI for acknowledging the processing of the spooled segment.
+        headers (property): The headers associated with the spooled segment.
+
+    Methods:
+        acknowledge(): Sends an acknowledgment request for the segment.
+    """
+    def __init__(
+        self,
+        response: _SpooledProtocolResponseTO,
+        segment: _SpooledSegmentTO,
+        row_mapper: RowMapper,
+        request: TrinoRequest
+    ) -> None:
+        super().__init__(segment)
+        self._segment = cast(_SpooledSegmentTO, segment)
+        self._row_mapper = row_mapper
+        self._request = request
+        self._encoding = response["encoding"]
+        self._headers = segment.get("headers")
+
+    @property
+    def rows(self) -> List[List[Any]]:
+        return self._row_mapper.map(json.loads(self._load_spooled_segment()))
+
+    @property
+    def uri(self) -> str:
+        return self._segment["uri"]
+
+    @property
+    def ack_uri(self) -> str:
+        return self._segment["ackUri"]
+
+    @property
+    def headers(self) -> Dict[str, List[str]]:
+        return self._segment.get("headers", {})
+
+    def acknowledge(self):
+        http_response = self._send_spooling_request(self._segment["ackUri"], self._segment)
+        if not http_response.ok:
+            self._request.raise_response_error(http_response)
+
+    def _load_spooled_segment(self) -> str:
+        segment = self._segment
+        http_response = self._send_spooling_request(segment["uri"], segment)
+        if not http_response.ok:
+            self._request.raise_response_error(http_response)
+        content = http_response.content
+
+        encoding = self._encoding
+        if encoding == "json+zstd":
+            zstd_decompressor = zstandard.ZstdDecompressor()
+            return zstd_decompressor.decompress(content).decode('utf-8')
+        elif encoding == "json+lz4":
+            expected_size = segment["metadata"]["uncompressedSize"]
+            return lz4.block.decompress(content, uncompressed_size=int(expected_size)).decode('utf-8')
+        elif encoding == "json":
+            return content.decode('utf-8')
+        else:
+            raise ValueError(f"Unsupported encoding: {encoding}")
+
+    def _send_spooling_request(self, uri: str, segment: _SpooledSegmentTO) -> requests.Response:
+        req = requests.Request(
+            "GET",
+            uri,
+            headers=list(map(lambda item: [(item[0], value) for value in item[1]], segment.get("headers", {}).items()))
+        )
+        prepared = self._request._prepare_request(req)
+        return self._request._send(prepared)
+
+    def __repr__(self):
+        return (
+            f"SpooledSegment(metadata={self.metadata}, uri={self.uri}), "
```

Comment:
do not log `uri` and `ackUri` and `headers` - these are sensitive and we should avoid surfacing these